File: programs/churn/churn.ml
(* churn: thread a six-field record through N LCG-fed steps, two fields updated
   per step. OCaml's records are immutable and `{ r with … }` is the obvious
   functional spelling, so this column allocates a fresh 7-word block per step
   on the minor heap — the nursery priced on record churn, against the
   contestants that mutate and the one that reuses. *)
type churn = { a : int; b : int; c : int; d : int; e : int; f : int }

let p_mod = 1000003

let rec walk i p x r =
  if i = 0 then r
  else
    let y = 75 * x mod 65537 in
    match p with
    | 0 -> walk (i - 1) 1 y { r with a = (r.c + r.e + y) mod p_mod; b = (r.d + r.f + y) mod p_mod }
    | 1 -> walk (i - 1) 2 y { r with c = (r.e + r.a + y) mod p_mod; d = (r.f + r.b + y) mod p_mod }
    | _ -> walk (i - 1) 0 y { r with e = (r.a + r.c + y) mod p_mod; f = (r.b + r.d + y) mod p_mod }

let () =
  let n = int_of_string Sys.argv.(1) in
  let r = walk n 0 ((n + 1) mod 65537) { a = 1; b = 2; c = 3; d = 4; e = 5; f = 6 } in
  Printf.printf "%d\n" r.a

File: programs/churn/churn.ts
// churn: thread a six-field record through N LCG-fed steps, two fields updated
// per step. AssemblyScript -> wasm (its own GC), on wasmtime.
//
// The imperative spelling mutates class fields in place, so the single Churn
// object is the only allocation in the program.
//
// The @assemblyscript/wasi-shim patches the built-in `process`/`console` (no
// imports); top-level statements compile into the WASI `_start`. Compiled with
// cwd at the shim's install dir so its `lib: ./assembly` glob resolves.
const P: u64 = 1000003;

class Churn {
  a: u64;
  b: u64;
  c: u64;
  d: u64;
  e: u64;
  f: u64;
  constructor() {
    this.a = 1;
    this.b = 2;
    this.c = 3;
    this.d = 4;
    this.e = 5;
    this.f = 6;
  }
}

// argv[0] is the module path; the value we passed is the last entry.
const n = i32(parseInt(process.argv[process.argv.length - 1], 10));
let x: u64 = u64((n + 1) % 65537);
const r = new Churn();
let p = 0;
for (let i = 0; i < n; i++) {
  x = (75 * x) % 65537;
  if (p == 0) {
    r.a = (r.c + r.e + x) % P;
    r.b = (r.d + r.f + x) % P;
    p = 1;
  } else if (p == 1) {
    r.c = (r.e + r.a + x) % P;
    r.d = (r.f + r.b + x) % P;
    p = 2;
  } else {
    r.e = (r.a + r.c + x) % P;
    r.f = (r.b + r.d + x) % P;
    p = 0;
  }
}
console.log(r.a.toString());
